firedrake/preconditioners/hypre_ams.py

```diff
@@ -0,0 +1,77 @@
+from firedrake.petsc import PETSc
+from firedrake import FunctionSpace, Constant, project, Interpolator, grad, TestFunction
+from firedrake.assemble import allocate_matrix, create_assembly_callable
+
+__all__ = ("HypreAMS")
+
+class HypreAMS(firedrake.PCSNESBase):
+    def initialize(self, obj):
+        if isinstance(obj, PETSc.PC):
+            A, P = obj.getOperators()
+        elif isinstance(obj, PETSc.SNES):
+            A, P = obj.ksp.pc.getOperators()
+        else:
+            raise ValueError("Not a PC or SNES?")
+
+        prefix = obj.getOptionsPrefix()
+        appctx = self.get_appctx(obj) 
+
+        f = appctx['state']
+        element = str(f.function_space().ufl_element().family()) #not sure this is right
+        mesh = f.function_space().mesh()
+        V = FunctionSpace(mesh, element, 1)
+
+        print("element:", element)
+        print("mesh:", mesh)
+        print("f:", f)
+        print("A:", A.getValuesCSR())
+        print("P:", P.getValuesCSR())
+
+        # build gradient matrix G
+        P1 = FunctionSpace(mesh, "Lagrange", 1)
+        Q = FunctionSpace(mesh, "N1curl", 1)
+        G = Interpolator(grad(TestFunction(P1)), Q).callable().handle

```

Comment:
`Q` should really be `V`. And then I think for general (not just linear) order, one wants `Pk = FunctionSpace(mesh, "Lagrange", V.ufl_element().degree())`